test-helpers/src/mpc_network/mock_with_delay.rs

```diff
@@ -0,0 +1,216 @@
+    party_id: PartyId,
+    /// The underlying stream that mocks a real connection
+    ///
+    /// Contains a send timestamp as well as the actual message
+    mock_conn: UnboundedDuplexStream<NetworkOutboundWithTimestamp>,
+    /// The buffered message on the receive side
+    recv_buffer: Option<NetworkOutboundWithTimestamp>,
+    /// The delay to introduce on each message
+    delay: Duration,
+    /// The current timer allocated for the delay, must be held so that it
+    /// remains registered with the runtime's timer driver
+    timer: Option<Pin<Box<tokio::time::Sleep>>>,
+}
+
+impl MockNetworkWithDelay {
+    /// Constructor
+    pub fn new(
+        party_id: PartyId,
+        stream: UnboundedDuplexStream<NetworkOutboundWithTimestamp>,
+        delay: Duration,
+    ) -> Self {
+        Self {
+            party_id,
+            mock_conn: stream,
+            recv_buffer: None,
+            delay,
+            timer: None,
+        }
+    }
+
+    /// Register a timer to awake after the delay
+    ///
+    /// This requires both creating a `tokio::time::Sleep` instance and then
+    /// polling it so that the timer is registered with the driver
+    #[allow(unused_must_use)]
+    pub fn register_timer(mut self: Pin<&mut Self>, duration: Duration, cx: &mut Context<'_>) {
+        assert!(!duration.is_zero(), "timer duration already elapsed");
+
+        // Create a timer and poll it
+        let mut pinned_timer = Box::pin(tokio::time::sleep(duration));
+        pinned_timer.as_mut().poll(cx);
+
+        // Store the pinned timer so that it remains registered with the runtime
+        self.timer.replace(pinned_timer);
+    }
+}
+
+#[async_trait]
+impl MpcNetwork for MockNetworkWithDelay {
+    fn party_id(&self) -> PartyId {
+        self.party_id
+    }
+
+    async fn close(&mut self) -> Result<(), MpcNetworkError> {
+        Ok(())
+    }
+}
+
+impl Stream for MockNetworkWithDelay {
+    type Item = Result<NetworkOutbound, MpcNetworkError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Check the buffer to see if a message is ready to be received
+        let (instant, message) =
+            if let Some((send_time, message)) = self.as_mut().recv_buffer.take() {
+                (send_time, message)
+            } else {
+                ready!(Box::pin(self.mock_conn.recv()).poll_unpin(cx))
+            };
+
+        // If the delay between the send time and now has elapsed, return the message
+        // Otherwise, buffer the message and register a timer to wake up after the delay
+        let duration_since_send = Instant::now() - instant;
+        if duration_since_send >= self.delay {
+            Poll::Ready(Some(Ok(message)))
+        } else {
+            self.recv_buffer = Some((instant, message));
+
+            let timer_delay = self.delay - duration_since_send;
+            self.register_timer(timer_delay, cx);
+            Poll::Pending
+        }
+    }
```

Comment:
I defer to your expertise here, but do we ever actually rely on the `timer` object here? Seems like the "delay" functionality actually comes from the fact that we return `Poll::Pending` if `duration_since_send >= self.delay`, and not from ever `.await`ing the timer (we only poll it once)

Reply:
Spoke in person, copying here for transparency:

We effectively defer the context management to the timer. Although we don't directly return anything from the `Sleep::poll` call, this call will internally register the context to be re-polled when the timer deadline is met.